src/Orleans/Lifecycle/LifecycleObservable.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans
+{
+    public class LifecycleObservable<TStage> : ILifecycleObservable<TStage>, ILifecycleObserver
+    {
+        private readonly ConcurrentDictionary<object, OrderedObserver> subscribers;
```

Comment:
TStage are the stages of the lifecycle.  The key in this dictionary is just an opaque id used to removed the item.  They are orthogonal.